docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+	// in this case, we are using the same listener for each Store
+	for _, key := range storeKeys {
+		listeners[key] = listener
+	}
+	// check that the writeDir exists and is writeable so that we can catch the error here at initialization if it is not
+	// we don't open a dstFile until we receive our first ABCI message
+	if err := fileutil.IsDirWriteable(writeDir); err != nil {
+		return nil, err
+	}
+	return &FileStreamingService{
+		listeners: listener
+		srcChan: listenChan,
+		filePrefix: filePrefix,
+		writeDir: writeDir,
+		marshaler: m,
+		stateCache: make([][]byte, 0),
+		cacheLock: new(sync.Mutex),
+	}, nil
+}
+
+// Listeners returns the StreamingService's underlying WriteListeners, use for registering them with the BaseApp
+func (fss *FileStreamingService) Listeners() map[sdk.StoreKey][]storeTypes.WriteListener {
+	return fss.listeners
+}
+
+func (fss *FileStreamingService) ListenBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) {
+	// create a new file with the req info according to naming schema
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+func (fss *FileStreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) {
+	// create a new file with the req info according to naming schema
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+func (fss *FileStreamingService) ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) {
+	// create a new file with the req info according to naming schema
+	// NOTE: if the tx failed, handle accordingly
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
+func (fss *FileStreamingService) Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-quitChan:
+				return
+                        case by := <-fss.srcChan:
+                        	append(fss.stateCache, by)
+			}
+		}
+	}()
+}
+```
+
+Writing to a file is the simplest approach for streaming the data out to consumers.
+This approach also provide the advantages of being persistent and durable, and the files can be read directly,
+or an auxiliary streaming services can read from the files and serve the data over a remote interface.
+
+#### Auxiliary streaming service
+
+We will create a separate standalone process that reads and internally queues the state as it is written out to these files
+and serves the data over a gRPC API. This API will allow filtering of requested data, e.g. by block number, block/tx hash, ABCI message type,
```

Comment:
grpc allows streaming responses.
It would be nice to allow subscriptions via such.

Not just searching for when changes happened to key X, but getting an event each time that happens (with the context: DeliverTx/Begin/End block)

Reply:
Completely agree, adding language to include subscriptions!